Adds element deferring for KSP to Room Processing

gfreivasc opened the pull request:
## Proposed Changes

Creates a map of `XTypeElement` to their originating `KSAnnotated` that can
be used to return elements deferred on `XProcessingStep` back to the KSP
`SymbolProcessor`.

## Testing

```
Test: ./gradlew \
    test connectedCheck \
    -x :room:room-benchmark:cC \
    -x :room:integration-tests:room-incremental-annotation-processing:test
```

## Issues Fixed

Fixes: [b/182822821](https://issuetracker.google.com/182822821)

gfreivasc committed:
Adds element deferring for KSP

Creates a map of `XTypeElement` to their originating `KSAnnotated` that can
be used to return elements deferred on `XProcessingStep` back to the KSP
`SymbolProcessor`.

Test: ./gradlew \
    test connectedCheck \
    -x :room:room-benchmark:cC \
    -x \
:room:integration-tests:room-incremental-annotation-processing:test
Fixes: b/182822821

yigit commented on room/compiler-processing/src/main/java/androidx/room/compiler/processing/XProcessingStep.kt:
```diff
@@ -66,18 +67,22 @@ interface XProcessingStep {
             ).filterIsInstance<KSClassDeclaration>()
                 .map {
-                    env.requireTypeElement(it.qualifiedName!!.asString())
+                    val element = env.requireTypeElement(it.qualifiedName!!.asString())
+                    elementMap[element] = it
+                    element
```
you could also cast this to KspTypeElement to get the declaration out of it but this soolution is fine too.

gfreivasc replied:
I think it's more cohesive to do that, indeed

yigit reviewed (approved):
Thank you!

yigit commented:
btw we need to update room as well but i can also do it as a followup.

https://github.com/androidx/androidx/blob/androidx-main/room/compiler/src/main/kotlin/androidx/room/RoomKspProcessor.kt#L56

gfreivasc committed:
Changes implementation of KSP type wiring

Instead of externally mapping XTypeElement instances to their
originating KSAnnotated, cast them to KspTypeElement that already
contains an explicit reference to the desired instance.

Also modifies RoomKspProcessor to leverage the API changes

gfreivasc commented:
Changed the implementation to use the reference that `KspTypeElement` holds. also changes KspRoomProcessor to make use of the API changes.

dlam commented:
Build should be passing again now - if you rebase this checks should pass.
